### ISSUE TITLE
Getting a valid value from StreamBuilder<DocumentSnapshot> into a child StreamBuilder and changing a widget

### ISSUE THREAD

**Issue by davidjeffery** [FirebaseExtended/flutterfire#2726]
I am trying to get a specific icon to appear based on whether the user has taken the survey or not.

At the moment, I am using `StreamBuilder` to listen for a given value in a document, which returns the survey name. I then want to use the survey name in the next `StreamBuilder`, which will look in a given Firestore collection (which is made up of the survey name followed by _entrants - so, for example, Survey_entrants) for a completed survey document, which will have the title of the user's unique id (named userid).

The problem I have now is that `surveyName` does return the name of the survey put in Cloud Firestore, and updates it when I change the value (I can see this by the commented-out `return new Text('$surveyName');` command). However, it does not seem to be passing that value into the next `StreamBuilder` - regardless of what I put in as the survey name, I get the check icon showing, suggesting `(snapshot1.hasData)` - even when that document does not exist.

I know the `surveyName` variable is working, but if I do `snapshot1.toString()` I get the error `Snapshot(ConnectionState.active, Instance of 'DocumentSnapshot', null)`. This must count has having data, hence showing the survey being taken. How do I correct this?

My code:

```
Positioned(
  right: 30,
  top: 20,
  child: StreamBuilder<DocumentSnapshot>(
    stream: Firestore.instance
    .collection('Controller')
    .document('Current Survey')
    .snapshots(),
    builder: (BuildContext context,
              AsyncSnapshot<DocumentSnapshot> snapshot) {
      if (!snapshot.hasData) {
        return CircularProgressIndicator();
      } else {
        var sn = snapshot.data;
        surveyName = sn["cs"];
//      return new Text('$surveyName');
        return StreamBuilder(
          stream: Firestore.instance
          .collection('$surveyName' + '_entrants')
          .document(userid)
          .snapshots(),
          builder: (BuildContext context, snapshot1) {
            if (!snapshot1.hasData) {
              return Icon(
                Foundation.burst_new,
                size: 48,
                color: Color(0xff303841),
              );
            } else if (snapshot1.hasData) {
              return Icon(
                Foundation.check,
                size: 48,
                color: Color(0xff303841),
              );
            } else {
              return Icon(
                MaterialIcons.error_outline,
                size: 48,
                color: Color(0xff303841),
              );
            }
          });
      }
    })),
```
**Flutter doctor**

```
C:\Users\djeffery\AndroidStudioProjects\rewardpolling>flutter doctor -v
[√] Flutter (Channel stable, v1.17.2, on Microsoft Windows [Version 10.0.17763.1039], locale en-GB)
    • Flutter version 1.17.2 at C:\App\flutter
    • Framework revision 5f21edf8b6 (11 days ago), 2020-05-28 12:44:12 -0700
    • Engine revision b851c71829
    • Dart version 2.8.3


[!] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at C:\Users\djeffery\AppData\Local\Android\sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[√] Android Studio (version 4.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 10 (API 29) (emulator)

! Doctor found issues in 1 category.
```

**Comment by TahaTesser** [comment]
Hi @davidjeffery  
Can you please provide  and a minimal complete reproducible code sample and dummy firestore data screenshot 
Thank you

**Comment by davidjeffery** [comment]
@TahaTesser no problem - this is my first time doing this so I hope I've done it correctly. 

```
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:rewardpolling/constants.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:rewardpolling/PostAuthPages/latest_survey.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter_icons/flutter_icons.dart';

FirebaseUser loggedInUser;

class Dashboard extends StatefulWidget {
  static const String id = 'dashboard';
  @override
  _DashboardState createState() => _DashboardState();
}

class _DashboardState extends State<Dashboard> {
  final _auth = FirebaseAuth.instance;
  var userid;
  String sn;
  String sn2;
  String surveyName;
  bool takenSurvey;
  String title;

  @override
  void initState() {
    super.initState();
    getCurrentUser();
    getUserID();
  }

  void getCurrentUser() async {
    try {
      final user = await _auth.currentUser();
      if (user != null) {
        loggedInUser = user;
      }
    } catch (e) {
      print(e);
    }
  }

  void getUserID() async {
    try {
      final FirebaseUser user = await _auth.currentUser();
      if (user != null) {
        userid = user.uid;
      }
    } catch (_getUserIDError) {
      print(_getUserIDError);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Color(0xffEEEEEE),
      body: SingleChildScrollView(
        child: Container(
          padding: EdgeInsets.symmetric(horizontal: 24.0),
          child: Column(
              mainAxisAlignment: MainAxisAlignment.start,
              crossAxisAlignment: CrossAxisAlignment.stretch,
              children: <Widget>[
                Stack(
                  alignment: Alignment.center,
                  children: <Widget>[
                    Container(
                      width: MediaQuery.of(context).size.width * 0.9,
                      child: SquareButton(
                          buttonTitle: "",
                          fontSize: 36,
                          onPressed: () async {
                            final snapShot = await Firestore.instance
                                .collection('$surveyName' + '_entrants')
                                .document(userid)
                                .get();
                            if (snapShot.exists) {
                              print('ERROR');
                            } else if (surveyName.isEmpty)
                              print('ERROR2');
                            else {
                              return Navigator.push(
                                  context,
                                  new MaterialPageRoute(
                                      builder: (BuildContext context) =>
                                          new TestSurvey(surveyName)));
                            }
                          }),
                    ),
                    Positioned(
                        right: 45,
                        top: 15,
                        child: StreamBuilder<DocumentSnapshot>(
                            stream: Firestore.instance
                                .collection('Controller')
                                .document('Current Survey')
                                .snapshots(),
                            builder: (BuildContext context,
                                AsyncSnapshot<DocumentSnapshot> snapshot) {
                              if (!snapshot.hasData) {
                                return CircularProgressIndicator();
                              } else {
                                var sn = snapshot.data;
                                surveyName = sn["cs"];
//                                return new Text('$surveyName');
                                return StreamBuilder<DocumentSnapshot>(
                                    stream: Firestore.instance
                                        .collection(surveyName + '_entrants')
                                        .document(userid)
                                        .snapshots(),
                                    builder: (BuildContext context, snapshot1) {
                                      if (!snapshot1.hasData) {
                                        return Icon(
                                          Foundation.burst_new,
                                          size: 36,
                                          color: Color(0xff303841),
                                        );
                                      } else if (snapshot1.hasData) {
                                        return Text(snapshot.toString(),
                                            overflow: TextOverflow.visible,
                                            style: TextStyle(fontSize: 8));
                                        return Icon(
                                          Foundation.check,
                                          size: 36,
                                          color: Color(0xff303841),
                                        );
                                      } else {
                                        return Icon(
                                          MaterialIcons.error_outline,
                                          size: 36,
                                          color: Color(0xff303841),
                                        );
                                      }
                                    });
                              }
                            })),
                    Container(
                      child: Text('Latest Survey',
                          style: TextStyle(
                              color: Colors.white,
                              fontSize: 36,
                              fontFamily: "Roboto")),
                    ),
                    Container(
                      width: MediaQuery.of(context).size.width * 0.9,
                      child: Opacity(
                        opacity: 0.0,
                        child: MaterialButton(
                            height: 75.0,
                            color: Color(0x00000000),
                            onPressed: () async {
                              final snapShot = await Firestore.instance
                                  .collection('$surveyName' + '_entrants')
                                  .document(userid)
                                  .get();
                              if (snapShot.exists) {
                                print('ERROR');
                              } else if (surveyName.isEmpty)
                                print('ERROR2');
                              else {
                                return Navigator.push(
                                    context,
                                    new MaterialPageRoute(
                                        builder: (BuildContext context) =>
                                            new TestSurvey(surveyName)));
                              }
                            }),
                      ),
                    ),
                  ],
                ),
              ]),
        ),
      ),
    );
  }
}
```
Link to images: https://imgur.com/a/AJYjcYu

The first image shows where the current survey info is kept 'cs', and the second shows the document I'm looking for, to see if it exists (the field is autogenerated by Firebase as a userID)

**Comment by TahaTesser** [comment]
Simplified code sample

<details> 
<summary>Code Sample</summary> 

``` 
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      home: Dashboard(),
    );
  }
}

class Dashboard extends StatefulWidget {
  @override
  _DashboardState createState() => _DashboardState();
}

class _DashboardState extends State<Dashboard> {
  final userid = "uHNc3gy4RGo9zlJeno8p";

  String surveyName = "NEW";

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Scaffold(
        backgroundColor: Color(0xffEEEEEE),
        body: StreamBuilder<DocumentSnapshot>(
            stream: Firestore.instance
                .collection('Controller')
                .document('Current Survey')
                .snapshots(),
            builder: (BuildContext context,
                AsyncSnapshot<DocumentSnapshot> snapshot) {
              if (!snapshot.hasData) {
                return CircularProgressIndicator();
              } else {
                return StreamBuilder<DocumentSnapshot>(
                    stream: Firestore.instance
                        .collection(surveyName + '_entrants')
                        .document(userid)
                        .snapshots(),
                    builder: (BuildContext context, snapshot1) {
                      if (!snapshot1.hasData) {
                        return Text('NO ENTRANTS');
                      } else if (snapshot1.hasData) {
                        return Text(snapshot1.data.data.toString(),
                            overflow: TextOverflow.visible,
                            style: TextStyle(fontSize: 25));
                      } else {
                        return Icon(Icons.add_alarm);
                      }
                    });
              }
            }),
      ),
    );
  }
}

``` 

</details> 

<details> 
<summary>flutter doctor -v</summary> 

``` 
[✓] Flutter (Channel dev, 1.19.0-4.0.pre, on Mac OS X 10.15.5 19F101, locale en-GB)
    • Flutter version 1.19.0-4.0.pre at /Users/tahatesser/Code/flutter_dev
    • Framework revision 2f7a59a8da (4 days ago), 2020-06-05 03:44:02 -0700
    • Engine revision d17c84e7af
    • Dart version 2.9.0 (build 2.9.0-13.0.dev 02915ec5ce)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.45.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

[✓] Connected device (4 available)
    • SM M305F   • 32003c30dc19668f • android-arm64  • Android 10 (API 29)
    • macOS      • macOS            • darwin-x64     • Mac OS X 10.15.5 19F101
    • Web Server • web-server       • web-javascript • Flutter Tools
    • Chrome     • chrome           • web-javascript • Google Chrome 83.0.4103.97

• No issues found!

``` 

</details>

**Comment by Salakar** [comment]
Hey, as part of our on-going work for #2582, this should have been resolved in our Firebase Firestore rework (#2913), there was some underlying issues around errors being swallowed in streams, which in this case may be the issue and not highlighting what the actual underlying issue is.

This rework has now been merged into master. We'll look at publishing some prereleases in the next few days. Thank you
